fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/services/MembershipServiceImpl.java

```diff
@@ -0,0 +1,398 @@
+        return containerResc.getPropertyResourceValue(RdfLexicon.MEMBERSHIP_RESOURCE);
+    }
+
+    private FedoraResource getFedoraResource(final String txId, final FedoraId fedoraId) {
+        try {
+            return resourceFactory.getResource(txId, fedoraId);
+        } catch (final PathNotFoundException e) {
+            throw new PathNotFoundRuntimeException(e);
+        }
+    }
+
+    private FedoraResource getParentResource(final FedoraResource resc) {
+        try {
+            return resc.getParent();
+        } catch (final PathNotFoundException e) {
+            throw new PathNotFoundRuntimeException(e);
+        }
+    }
+
+    @Override
+    public void resourceDeleted(final String txId, final FedoraId fedoraId) {
+        // delete DirectContainer, end all membership for that source
+        FedoraResource fedoraResc;
+        try {
+            fedoraResc = getFedoraResource(txId, fedoraId);
+        } catch (final PathNotFoundRuntimeException e) {
+            log.debug("Deleted resource {} does not have a tombstone, cleanup any references");
+            indexManager.deleteMembershipReferences(txId, fedoraId);
+            return;
+        }
+        if (fedoraResc instanceof Tombstone) {
+            fedoraResc = ((Tombstone) fedoraResc).getDeletedObject();
+        }
+
+        if (isDirectContainer(fedoraResc)) {
+            indexManager.endMembershipForSource(txId, fedoraId, fedoraResc.getLastModifiedDate());
+        }
+
+        // delete child of DirectContainer, clear from tx and end existing
+        final var parentResc = getParentResource(fedoraResc);
+
+        if (isDirectContainer(parentResc)) {
+            final var parentRdfResc = getRdfResource(parentResc);
+            final var deletedMembership = generateDirectMembership(txId, parentRdfResc, fedoraResc);
+            indexManager.endMembership(txId, parentResc.getFedoraId(), deletedMembership,
+                    fedoraResc.getLastModifiedDate());
+
+        }
+    }
+
+    @Override
+    public RdfStream getMembership(final String txId, final FedoraId fedoraId) {
+        final FedoraId subjectId;
+        if (fedoraId.isDescription()) {
+            subjectId = fedoraId.asBaseId();
+        } else {
+            subjectId = fedoraId;
+        }
+        final var subject = NodeFactory.createURI(subjectId.getBaseId());
+        final var membershipStream = indexManager.getMembership(txId, subjectId);
+        return new DefaultRdfStream(subject, membershipStream);
+    }
+
+    @Override
+    public void commitTransaction(final String txId) {
+        indexManager.commitTransaction(txId);
+    }
+
+    @Override
+    public void rollbackTransaction(final String txId) {
+        indexManager.deleteTransaction(txId);
+    }
+
+    @Override
+    public void reset() {
+        indexManager.clearIndex();
+    }
+
+    @Override
+    public void populateMembershipHistory(final String txId, final FedoraId containerId) {
```

Comment:
`@Transactional`

Reply:
Sure. In the case of a container with lots of members this could be a very long transaction, I'm guessing that won't be problematic?

Reply:
No, I wouldn't expect that would be a problem